### ISSUE TITLE
symbols are not idempotent

### ISSUE THREAD

**Issue by primo-ppcg** [janet-lang/janet#1192]
Shouldn't they be?

In the "Prepare for 1.27.0" commit, `idempotent?` was refactored, with `:symbol` being added to the list of `non-atomic-types`:

https://github.com/janet-lang/janet/commit/01aab666676c5f93529ba8db5d97b9db1f4c499a#diff-e120880268b4f0f04177470180f50ee0d2c7ac13cb83bb778b6d81efda1cbbcc

I think this may have been inadvertent. Throughout the core library, this function is used to determine if a gensym is necessary to prevent a form from being evaluated twice.

As a result, we have more complicated expansions than necessary:

```janet
(def a 1)
(def b 2)
(macex ~(or a b))
# ->
(do (def _00000r a) (if _00000r _00000r b))
```
where
```janet
(if a a b)
```
would clearly be sufficient.

**Comment by bakpakin** [comment]
So yes, the issue is fundamentally  that symbols are only idempotent in certain contexts (we know that the symbol is not redefined or mutated between all evaluations of the symbol)

We probably could have a separate (possibly internal) macro with such a constraint and use it there.

**Comment by primo-ppcg** [comment]
> Currently, this prints `0123456789`. If I reimplement `loop` so that symbols are considered idempotent, it prints `0259`.

I've confirmed that this was the behavior in 1.26.0, that is, until just a few months ago. One or the other is not obviously more correct to me.

**Comment by primo-ppcg** [comment]
`idempotent?` is used to answer several different questions throughout the core library.

As it relates to symbols specifically:
1. Q: Can this appear multiple times in an expansion without potentially causing side-effects? (which includes `and`, `or`, `case`, `match`, `each` and any other place where these are used)
A: Yes.
2. Q: Does this represent an atomic binding target? (which includes `if-let`, `when-let`, `if-with`, `when-with` and any other place where these are used)
A: Yes.
3. Q: Is this known not to change throughout subsequent iterations of a user provided body? (which very specifically applies to the `step` argument of a range iterator, and nowhere else as far as I have been able to determine)
A: No.

For case 2., `symbol?` could be used instead (in fact [I've done so as a work-around](https://github.com/janet-lang/janet/pull/1190/files#diff-e120880268b4f0f04177470180f50ee0d2c7ac13cb83bb778b6d81efda1cbbccR678)). For case 3., `number?` could be used instead if it's really a concern.

For case 1., which is by far the most common usage of `idempotent?`, there is no alternative, other than to define a separate function.

Futhermore, the definition of the function is as given in the documentation, not more or less.
[idempotent?](https://janet-lang.org/api/index.html#idempotent?)
_Check if x is a value that evaluates to itself when compiled._
```janet
(setdyn *pretty-format* "%m")
(pp (disasm (fn [] 'a)))
# ->
{ :arity 0
  :bytecode @[(ldc 0 0) (ret 0)]
  :constants @[a]
  :defs @[]
  :environments @[]
  :max-arity 0
  :min-arity 0
  :slotcount 1
  :source "sym.janet"
  :sourcemap @[(2 14) (2 14)]
  :structarg false
  :vararg false}
```
And finally, I'm not convinced that he behavior of 1.26.0 regarding the `:range` iterator is wrong or undesirable. If the `step` argument is modified within the body, it almost certainly would have been done with intent. For example:
```janet
(var b 1)
(seq [a :range [1 200 b]]
  (set b (- a b)))
# ->
@[0 1 1 2 3 5 8 13 21 34 55 89]
```
If we're really caught up on the name, because Webster or whoever disagrees with the definition given in the documentation, we could define a second function, and then just never use the original anywhere. I don't see any practical value in that, though.

**Comment by sogaiu** [comment]
Thanks for the detailed explanation.

I did some digging into the past...

IIUC, there used to be an ancestor(?) function named `atomic?` and this was changed [here](https://github.com/janet-lang/janet/commit/b250679789a3f002a20de138199a7678344b8b7c#diff-1a364b6b8ae784fe1510a644222ded9c864bba2b7db51d210e25b98477363925).

Although I do find the behavior slightly surprising given the current name, the intent of "when compiled" seems to have been there for quite some time -- in fact it seems to have been [in the very first docstring](https://github.com/janet-lang/janet/commit/4d119e4e03bffb2c83e208e9be7bf26e01a127db#diff-1a364b6b8ae784fe1510a644222ded9c864bba2b7db51d210e25b98477363925R126).

I think it's often difficult to come up with an appropriate short name for something that adequately captures nuance and if one has reflected on naming frequently enough this becomes something one is likely to eventually accept :)

For internal use it's probably clearer how something ought to behave here, but perhaps it's less clear how a public function should behave (in this case there was a history of acting in a particular fashion).

In any case, I'm not sure it was used that much in folks' code -- at least a search across the Janet code I've collected locally doesn't suggest it was that widely used.

**Comment by primo-ppcg** [comment]
> So I am in favor of bakpakin's idea of:
> 
> > We probably could have a separate (possibly internal) macro with such a constraint and use it there.

We could go back to the original definition of `atomic?`.

```diff
diff --git a/src/boot/boot.janet b/src/boot/boot.janet
@@ -129,7 +129,7 @@
 # For macros, we define an imcomplete odd? function that will be overriden.
 (defn odd? [x] (= 1 (mod x 2)))
 
-(def- non-atomic-types
+(def- non-idempotent-types
   {:array true
    :tuple true
    :table true
@@ -140,4 +140,15 @@
 (defn idempotent?
   "Check if x is a value that evaluates to itself when compiled."
   [x]
+  (not (in non-idempotent-types (type x))))
+
+(def- non-atomic-types
+  {:array true
+   :tuple true
+   :table true
+   :struct true})
+
+(defn- atomic?
+  "Check if x is an atomic type."
+  [x]
   (not (in non-atomic-types (type x))))
```
... and then use that everywhere.

**Comment by primo-ppcg** [comment]
> It applies in each:

Yes, and this is also very useful. Appending to an array as you iterate it is a very common paradigm for implementing a breadth-first search. As an example, a Sudoku solver:

```janet
(def s @[`
3__5____4
49___8___
___4915__
__1____8_
___1___47
543___9_6
__6_4____
839275___
_7_61__59`])

(each x s
  (when-let [i (index-of 95 x)]
    (def row (seq [j :range [0 9]] (in x (+ (- i (mod i 10)) j))))
    (def col (seq [j :range [0 9]] (in x (+ (mod i 10) (* j 10)))))
    (def box (seq [j :range [0 9]]
      (in x (+ (- i (mod i 30) (mod (mod i 10) 3)) (mod i 10) (mod (* j 10) 29)))))
    (def used (distinct [;row ;col ;box]))
    (def chars (filter (fn [c] (= nil (index-of c used))) "123456789"))
    (each c chars
      (def y (buffer x))
      (set (y i) c)
      (array/push s y))))

(print (last s))
# ->
318527694
495368172
627491538
761934285
982156347
543782916
156849723
839275461
274613859
```
> I understand that my examples are contrived and edge cases

That I will agree with. Not to mention that what is happening entirely intuitive and _useful_. It is my opinion that the user can be trusted to manage their own variable references in a sensible way. The only reason to not allow it is because WeSaySo™, and by doing so have deliberately slowed your code besides. You're welcome.

> Their code would maybe be slightly slower, but that's it.

The difference is in fact much bigger than you realize. That's precisely why I raised the issue in the first place.

Using [this](https://github.com/janet-lang/spork/pull/128/commits/e0ac5f00898159dbdcdd9970914cdcfd814d5a70#diff-53ad22c96e1a69d0acc5b381cb89915d70a2dfb9e7efe5eff81cb7a798c109feR75-R102) profiling tool:
```janet
(def a 1)
(def b 2)
(timeit-loop [:repeat 500_000_000] (or a b))

master     6.026s, 0.01205µs/body
no-symbol  4.167s, 0.008335µs/body
```
That's 44% faster.

**Comment by primo-ppcg** [comment]
I'm also just expressing opinions. I'm not the one who makes decisions around here 😉

Really though, I don't see the benefit. To deliberately slow every each, every range iterator, every match, every case, because someone, somewhere, at some point it time might try to modify a variable - doesn't make much sense to me. And more likely than not, the aforementioned hypothetical person will have done so deliberately, and be mildly annoyed that it doesn't work. Perhaps I'm just too much of a pragmatist.

It's a fair point, though, that there is a reasonable expectation for functions in the public API to be correct in their implementation. That the adding of `:symbol` to the list of non-idempotent-type - which, I now agree was probably correct - slowed many dependent functions and macros I still believe was inadvertent. All of the so affected, of course, pre-date the 1.27 release.

**Comment by primo-ppcg** [comment]
> Possibly an even better way to get this speed back might be to add a compiler pass that detects unnecessary defs and removes them.

Great progress has already been made towards this, just a few weeks ago: #1163 

Testing methodology is also slightly different, I recompiled the current master with `:symbol` removed from the list of `non-atomic-types`. But I also found that a `var` binding benefits from being aliased. I wonder if this could somehow be exposed (and then we could be in complete agreement that var bindings are not idempotent).

> I made my peace with lisp being slow a long time ago

I used Lisp specifically because it was fast (SBCL).
